### PR TITLE
cleanup chains

### PR DESCRIPTION
**Pull request by bekzod**

### CODE REVIEW

File: packages/ember-metal/lib/chains.ts [hunk 1]
```diff
@@ -191,21 +190,13 @@ class ChainNode {
     this._paths = undefined;
 
     // It is false for the root of a chain (because we have no parent)
-    let isWatching = parent !== null;
-    let object;
-    if (isWatching) {
+    if ((this._isWatching = parent !== null)) {
```
Comment (locks) [inline packages/ember-metal/lib/chains.ts]: I'm not a fan of this pattern, it kinda seems like a mistake if you're reading with full attention.

File: packages/ember-metal/lib/chains.ts [hunk 2]
```diff
@@ -266,16 +258,14 @@ class ChainNode {
...
-    if (this._chains === undefined) {
+    let chains = this._chains;
+    if (chains === undefined) {
       chains = this._chains = Object.create(null);
-    } else {
-      chains = this._chains;
     }
 
-    let node = chains[key];
+    let node = chains![key];
```
Comment (krisselden) [inline packages/ember-metal/lib/chains.ts]: the `!` shouldn't be needed because chains is already checked for undefined and assigned.

Comment (bekzod) [reply packages/ember-metal/lib/chains.ts]: linter is complaining if I do that

**Review by krisselden — CHANGES_REQUESTED** [review]
the `!` chains shouldn't be needed

File: packages/ember-metal/lib/chains.ts [hunk 3]
```diff
@@ -191,21 +190,14 @@ class ChainNode {
...
 
     // It is false for the root of a chain (because we have no parent)
-    let isWatching = parent !== null;
-    let object;
+    let isWatching = (this._isWatching = parent !== null);
     if (isWatching) {
       let parentValue = parent!.value();
       if (isObject(parentValue)) {
-        object = parentValue;
+        this._object = parentValue;
```
[the remark this answers is not in the document]
Comment (bekzod) [reply packages/ember-metal/lib/chains.ts]: should I remove `this._object = undefined` then ?

**Comment by krisselden** [comment]
I believe this was superseded by #16680 if not, please rebase and reopen

**Comment by bekzod** [comment]
@krisselden rebased please reopen :P cannot do it myself

File: packages/ember-metal/lib/chains.ts [hunk 4]
```diff
@@ -256,16 +248,14 @@ class ChainNode {
...
-    if (this.chains === undefined) {
+    let chains = this.chains;
+    if (chains === undefined) {
       chains = this.chains = Object.create(null);
-    } else {
-      chains = this.chains;
     }
 
-    let node = chains[key];
+    let node = chains![key];
```
Comment (krisselden) [inline packages/ember-metal/lib/chains.ts]: I don't understand why the `!` is needed here.

Comment (krisselden) [reply packages/ember-metal/lib/chains.ts]: The `if (chains === undefined)` and assignment should remove undefined.

**Review by krisselden — CHANGES_REQUESTED** [review]

**Comment by bekzod** [comment]
I wouldn’t add it if linter was not complaining

File: packages/ember-metal/lib/chains.ts (hunk 4 above)
Comment (bekzod) [reply packages/ember-metal/lib/chains.ts]: Linter is complaining if I remove it

**Comment by bekzod** [comment]
any more comments on this ?

File: packages/ember-metal/lib/chains.ts (hunk 4 above)
Comment (bekzod) [reply packages/ember-metal/lib/chains.ts]: probably bug in tslinter

**Comment by bekzod** [comment]
@rwjblue I think this is good to go 😉

**Pull request merged**